src/qutip_tensorflow/core/data/add.py

```diff
@@ -0,0 +1,60 @@
+import qutip
+from .tftensor import TfTensor
+import warnings
+
+with warnings.catch_warnings():
+    warnings.filterwarnings("ignore", category=DeprecationWarning)
+    from tensorflow.errors import InvalidArgumentError
+
+__all__ = ["add_tftensor", "sub_tftensor", "iadd_tftensor"]
+
+def _check_shape(left, right):
+    if left.shape != right.shape:
+        raise ValueError(
+            f"""Incompatible shapes for addition of two matrices:
+                         left={left.shape} and right={right.shape}"""
+        )
```

Comment:
So I considered this at the time, and the reason I didn't do it immediately is because the core data types would always have to repeat the test anyway; if you access the raw function (say by `qutip.core.data.add_csr`, or whatever), then you'd be able to segfault Python by passing it bad input.  I also had vague thoughts that some data-layer types might want to eschew shape checks (at the time, @Ericgig was thinking about an adaptive Hilbert space type which automatically expanded or something), or we might want symbolic backing types which could represent shapes without integers.  Obviously those don't jive entirely with the definition of `data.Data`, but that's the sort of thing I had in mind.